File: 2°-Periodo/AEDs2/Exercícios/arvoreBinaria.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

typedef struct No
{
    int elemento;
    int *dir, *esq;
} No;

No *raiz;

void construtor()
{
    raiz = NULL;

    No no;
    no.elemento = NULL;
    no.dir = NULL;
    no.esq = NULL;
}

void inserir(int x)
{
    inserirRec(x, &raiz);
}

void inserirRec(int x, No **i)
{
    if (*i == NULL)
    { // Se tiver vazio, apenas insere
        *i = novoNo(x);
    }
    else if (x < (*i)->elemento)
    { // Se o elemento for menor, volta recursivo para a esquerda
        inserirRec(x, &((*i)->esq));
    }
    else if (x > (*i)->elemento)
    { // Se o elemento for maior, volta recursivo para a direita
        inserirRec(x, &((*i)->dir));
    }
    else
    {
        errx(1, "Erro ao inserir!");
    }
}

void caminharCentral()
{
    printf("[ ");
    caminharCentralRec(raiz);
    printf("]\n");
}

void caminharCentralRec(No *i)
{
    if (i != NULL)
    {
        caminharCentralRec(i->esq);
        printf("%d ", i->elemento);
        caminharCentralRec(i->dir);
    }
}

int main()
{
    int x = 0;
    while (x >= 0)
    {
        scanf("%d", &x);
        inserir(x);
    }
    mostrar();
}

File: 2°-Periodo/AEDs2/Estudos/ArvoreBinaria.java
import java.util.*;
class No{
    int elemento;
    int qtd;
    No dir, esq;

    public No(){
        this.elemento=0;
        this.qtd=0;
        this.dir=null;
        this.esq=null;
    }

    public No(int elemento){
        this.elemento=elemento;
        this.qtd=0;
        this.dir=null;
        this.esq=null;
    }

    public No(int elemento, No dir, No esq){
        this.elemento=elemento;
        this.qtd=0;
        this.dir=dir;
        this.esq=esq;
    }
}
public class ArvoreBinaria {
    No raiz;

    ArvoreBinaria(){
        raiz=null;
    }

    void inserir(int x){
        raiz = inserir(x, raiz);
    }

    No inserir(int x, No i){
        if(i==null){
            No tmp = new No(x);
        } else if(x>i.elemento){
            i.dir = inserir(x, i.dir);
        } else if(x<i.elemento){
            i.esq = inserir(x, i.esq);
        } else{
            i.qtd++;
        }
        return i;
    }

    void inserirPai(int x) throws Exception{
         if(raiz==null){
            No tmp = new No(x);
         } else if(x > raiz.elemento){
            inserirPai(x, raiz.dir, raiz);
         } else if(x < raiz.elemento){
            inserirPai(x, raizesq, raiz);
         } else{
            throw new Exception("ERRO");
         }
    }
}